restore/plan/plan.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 RetailNext, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"context"
+
+	"github.com/retailnext/cassandrabackup/bucket"
+	"github.com/retailnext/cassandrabackup/digest"
+	"github.com/retailnext/cassandrabackup/manifests"
+	"github.com/retailnext/cassandrabackup/unixtime"
+	"go.uber.org/zap"
+)
+
+type HistoryEntry struct {
+	Manifest manifests.ManifestKey
+	Digest   digest.ForRestore
+}
+
+type NodePlan struct {
+	Files             map[string]digest.ForRestore
+	ChangedFiles      map[string][]HistoryEntry
+	SelectedManifests manifests.ManifestKeys
+}
+
+func Create(ctx context.Context, identity manifests.NodeIdentity, startAfter, notAfter unixtime.Seconds) (NodePlan, error) {
+	lgr := zap.S().With("identity", identity)
+
+	nodeManifests, err := getManifests(ctx, identity, startAfter, notAfter)
+	if err != nil {
+		lgr.Errorw("get_manifests_error", "err", err)
+		return NodePlan{}, err
+	}
+
+	return assemble(nodeManifests), nil
+}
+
+func getManifests(ctx context.Context, identity manifests.NodeIdentity, startAfter, notAfter unixtime.Seconds) ([]manifests.Manifest, error) {
+	client := bucket.OpenShared()
+
+	keys, err := client.ListManifests(ctx, identity, startAfter, notAfter)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
you might also want to ensure here that `len(keys)>0`

Reply:
With an empty keys, the for loop won't do anything (i would be < 0 and the condition is `i >= 0`), and then it'll hit the explicit check before `client.GetManifests`.

Reply:
If keys is empty, the for loop body never executes (because `i >= 0` is never true) and we fall through to the explicit check on len before the `client.GetManifests` call.